src/openassetio-core-c/private/SimpleMap.cpp

```diff
@@ -0,0 +1,99 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright 2013-2022 The Foundry Visionmongers Ltd
+#include <stdexcept>
+
+#include <openassetio/c/SimpleMap.h>
+#include <openassetio/c/StringView.h>
+#include <openassetio/c/errors.h>
+#include <openassetio/SimpleMap.hpp>
+#include <openassetio/typedefs.hpp>
+
+#include "StringView.hpp"
+#include "chandle.hpp"
+#include "errors.hpp"
+
+namespace {
+
+using openassetio::SimpleMap;
+using HandleConverter = openassetio::chandle::Converter<SimpleMap, OPENASSETIO_NS(SimpleMap_h)>;
+
+/**
+ * Get a primitive value from a SimpleMap, converting exceptions to
+ * error codes.
+ *
+ * @tparam Type Type of value to extract from variant.
+ * @param[out] err Storage for error message, if any.
+ * @param[out] out Storage for return value.
+ * @param handle Opaque handle to a SimpleMap.
+ * @param key Key to query within SimpleMap.
+ * @return Error code.
+ */
+template <class Type>
+int get(OPENASSETIO_NS(StringView) * err, Type *out, OPENASSETIO_NS(SimpleMap_h) handle,
+        const OPENASSETIO_NS(ConstStringView) key) {
+  const SimpleMap *simpleMap = HandleConverter::toInstance(handle);
+
+  return openassetio::errors::catchUnknownExceptionAsCode(err, [&] {
+    // TODO(DF): @exception messages.
+    try {
+      *out = std::get<Type>(simpleMap->at({key.data, key.size}));
+    } catch (const std::out_of_range &exc) {
+      // Default exception message:
+      // VS 2019: "invalid unordered_map<K, T> key"
+      // GCC 9: "_Map_base::at"
+      openassetio::assignStringView(err, "Invalid key");
+      return OPENASSETIO_NS(kError_OutOfRange);
+    } catch (const std::bad_variant_access &exc) {
+      // Default exception message:
+      // VS 2019: "bad variant access"
+      // GCC 9: "Unexpected index"
+      openassetio::assignStringView(err, "Invalid value type");
+      return OPENASSETIO_NS(kError_BadVariantAccess);
+    }
+
+    return OPENASSETIO_NS(kOK);
+  });
+}
+}  // namespace
+
+extern "C" {
+
+OPENASSETIO_NS(SimpleMap_s) OPENASSETIO_NS(SimpleMap_suite)() {
+  return {// dtor
+          [](OPENASSETIO_NS(SimpleMap_h) handle) { delete HandleConverter::toInstance(handle); },
```

Comment:
Ouch. My brain hurts.

> This [user-defined conversion function](https://en.cppreference.com/w/cpp/language/cast_operator) is only defined if the capture list of the lambda-expression is empty. It is a public, constexpr, (since C++17) non-virtual, non-explicit, const noexcept member function of the closure object.

> The value returned by this conversion function is a pointer to a function with C++ [language linkage](https://en.cppreference.com/w/cpp/language/language_linkage) that, when invoked, has the same effect as invoking the closure object's function call operator directly.

This potentially means the actual function pointer for the lambda may have a different linkage (and therefor ABI) than a plain-C function pointer of the same parameters.

I think the safe way to do this would have to be declaring the function extern "C" with a C++ definition and taking a pointer to that.

Reply:
This is a super interesting point, which I've still not fully absorbed :brain:,  though when we deal with your other good point in  https://github.com/TheFoundryVisionmongers/OpenAssetIO/pull/330#discussion_r855645900 then the lambdas will likely disappear.

Reply:
Reading a little more on this, I believe the concern would be if the lambda function pointer's C++ language linkage implied a different calling convention than would be expected by a C caller.  From what I understand, this would never be the case for us, and would be difficult for anyone to cause.

Though I admit this is right at the edge of my level of understanding.  Always more to learn with C++!